linkerd/app/outbound/src/ingress.rs

```diff
@@ -1,45 +1,61 @@
 use crate::{http, stack_labels, tcp, trace_labels, Config, Outbound};
 use linkerd_app_core::{
     config::{ProxyConfig, ServerConfig},
-    detect, discovery_rejected, drain, errors, http_request_l5d_override_dst_name_addr,
-    http_tracing, io, profiles,
+    detect, discovery_rejected, errors, http_request_l5d_override_dst_name_addr, http_tracing, io,
+    profiles,
     proxy::api_resolve::Metadata,
     svc::{self, stack::Param},
     tls,
-    transport::{self, OrigDstAddr, Remote, ServerAddr},
+    transport::{OrigDstAddr, Remote, ServerAddr},
     AddrMatch, Conditional, Error, NameAddr,
 };
 use std::convert::TryFrom;
 use thiserror::Error;
 use tracing::{debug_span, info_span};
 
-impl Outbound<()> {
+#[derive(Clone)]
+struct AllowHttpProfile(AddrMatch);
+
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+struct Target {
+    dst: NameAddr,
+    version: http::Version,
+}
+
+#[derive(Clone)]
+struct TargetPerRequest(http::Accept);
+
+#[derive(Debug, Error)]
+#[error("ingress-mode routing requires a service profile")]
+struct ProfileRequired;
+
+#[derive(Debug, Error)]
+#[error("ingress-mode routing requires the l5d-dst-override header")]
+struct DstOverrideRequired;
+
+#[derive(Debug, Default, Error)]
+#[error("ingress-mode routing is HTTP-only")]
+struct IngressHttpOnly;
+
+// === impl Outbound ===
+
+impl<H> Outbound<H> {
     /// Routes HTTP requests according to the l5d-dst-override header.
     ///
     /// Forwards TCP connections without discovery/routing (or mTLS).
     ///
     /// This is only intended for Ingress configurations, where we assume all
     /// outbound traffic is either HTTP or TLS'd by the ingress proxy.
-    pub fn to_ingress<T, I, N, NSvc, H, HSvc, P>(
-        &self,
+    pub fn into_ingress<T, I, HSvc, P>(
```

Comment:
nit/TIOLI: maybe this should be called `into_http_ingress`, or something? just to make it clearer that it is now http only.